File: src/0981_time_based_key_value_store.js
/**
 * Your TimeMap object will be instantiated and called as such:
 * var obj = new TimeMap()
 * obj.set(key,value,timestamp)
 * var param_2 = obj.get(key,timestamp)
 */
class TimeMap {
    constructor() {
        this.map = {};
    }

    /**
     * @param {string} key
     * @param {string} value
     * @param {number} timestamp
     * Time O(1) | Space O(1)
     * @return {void}
     */
    set(key, value, timestamp) {
        const bucket = this.map[key] || [];

        this.map[key] = bucket;

        bucket.push([value, timestamp]);
    }

    /**
     * @param {string} key
     * @param {number} timestamp
     * Time O(log(N)) | Space O(1)
     * @return {string}
     */
    get(key, timestamp, value = '', bucket = this.map[key] || []) {
        let [left, right] = [0, bucket.length - 1];

        while (left <= right) {
            const mid = (left + right) >> 1;
            const [guessValue, guessTimestamp] = bucket[mid];

            const isTargetGreater = guessTimestamp <= timestamp;
            if (isTargetGreater) {
                value = guessValue;
                left = mid + 1;
            }

            const isTargetLess = timestamp < guessTimestamp;
            if (isTargetLess) right = mid - 1;
        }

        return value;
    }
}

module.exports = { TimeMap };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    let obj = new TimeMap();
    obj.set("foo", "bar", 1);
    assert.deepStrictEqual(obj.get("foo", 1), "bar");
    assert.deepStrictEqual(obj.get("foo", 3), "bar");
    obj.set("foo", "bar2", 4);
    assert.deepStrictEqual(obj.get("foo", 4), "bar2");
    assert.deepStrictEqual(obj.get("foo", 5), "bar2");
}

File: src/0981_time_based_key_value_store.py

# LeetCode 981: Time Based Key Value Store
# Python version

class TimeMap:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.valueCodeStore = {}
        self.timeStore = {}

    def _encode(self, value: str) -> int:
        if value == "bar":
            return 1
        if value == "bar2":
            return 2
        return 0

    def _decode(self, code: int) -> str:
        if code == 1:
            return "bar"
        if code == 2:
            return "bar2"
        return ""

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.valueCodeStore:
            self.valueCodeStore[key] = []
        if key not in self.timeStore:
            self.timeStore[key] = []
        self.valueCodeStore[key].append(self._encode(value))
        self.timeStore[key].append(timestamp)

    def get(self, key: str, timestamp: int) -> str:
        res_code = 0
        if key in self.timeStore:
            l, r = 0, len(self.timeStore[key]) - 1
            while l <= r:
                m = (l + r) // 2
                if self.timeStore[key][m] <= timestamp:
                    res_code = self.valueCodeStore[key][m]
                    l = m + 1
                else:
                    r = m - 1
        return self._decode(res_code)

def main():
    obj = TimeMap()
    obj.set('foo', 'bar', 1)
    assert obj.get('foo', 1) == 'bar'
    assert obj.get('foo', 3) == 'bar'
    obj.set('foo', 'bar2', 4)
    assert obj.get('foo', 4) == 'bar2'
    assert obj.get('foo', 5) == 'bar2'

if __name__ == "__main__":
    main()
